Translation of Chapter 4

WUTONK opened the pull request:
Translation of Chapter 4

WUTONK committed:
Update chapter4.md

Upload Volume 4 again

Qubik65536 reviewed (changes requested):
请把行数对齐，不需要的换行不要换行，原来的内容是第几行现在就还是第几行

WUTONK committed:
zh_cn chapter4

Unnecessary line breaks are removed

WUTONK commented:
已经修改

WUTONK committed:
zh_cn chapter4

The line break spacing is aligned

Qubik65536 reviewed (changes requested):
还得对齐

WUTONK commented:
对齐了

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
 # Rendering
 
-In this chapter we will learn the processes that takes place while rendering a scene using OpenGL. If you are used to older versions of OpenGL, that is fixed-function pipeline, you may end this chapter wondering why it needs to be so complex. You may end up thinking that drawing a simple shape to the screen should not require so many concepts and lines of code. Let me give you an advice for those ...
+在本章中,我们将学习使用OpenGL渲染场景时发生的过程和三维图形基本渲染过程.如果你习惯于旧版本的OpenGL,也就是固定的函数管道,你可能会在本章结束时想知道为什么它需要如此复杂.你可能会认为,只是在屏幕上画一个简单的形状完全不需要那么多的概念和代码.让我给你们这些有这种想法的人一个建议.它实际上更简单、更灵活.你只需要给它一个机会:现代的OpenGL让你一次只思考一个问题,它让你以一种更有逻辑的方式组织你的代码和进程.
```
> 固定的函数管道

`fixed-function pipeline`，固定渲染管线

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
+将一个三维表示绘制到二维屏幕的一系列步骤称为图形管道.OpenGL的第一个版本使用了一个称为固定函数管道的模型.该模型在渲染过程中采用了一组步骤,定义了一组固定的操作,程序员被限制在每个步骤可用的一组函数中.因此,可以应用的效果和操作受到API本身的限制\(例如，“set fog”或“add light”，但这些函数的实现是固定的，不能更改\).
 
-The graphics pipeline was composed of these steps:
+图形管道由以下步骤组成：
```
> 图形管道

`graphics pipeline`，渲染管线

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
+在本章中,我们将学习使用OpenGL渲染场景时发生的过程和三维图形基本渲染过程.如果你习惯于旧版本的OpenGL,也就是固定的函数管道,你可能会在本章结束时想知道为什么它需要如此复杂.你可能会认为,只是在屏幕上画一个简单的形状完全不需要那么多的概念和代码.让我给你们这些有这种想法的人一个建议.它实际上更简单、更灵活.你只需要给它一个机会:现代的OpenGL让你一次只思考一个问题,它让你以一种更有逻辑的方式组织你的代码和进程.
 
-The sequence of steps that ends up drawing a 3D representation into your 2D screen is called the graphics pipeline. First versions of OpenGL employed a model which was called fixed-function pipeline. This model employed a set of steps in the rendering process which defined a fixed set of operations. The programmer was constrained to the set of functions available for each step. Thus, the effects  ...
+将一个三维表示绘制到二维屏幕的一系列步骤称为图形管道.OpenGL的第一个版本使用了一个称为固定函数管道的模型.该模型在渲染过程中采用了一组步骤,定义了一组固定的操作,程序员被限制在每个步骤可用的一组函数中.因此,可以应用的效果和操作受到API本身的限制\(例如，“set fog”或“add light”，但这些函数的实现是固定的，不能更改\).
```
> 固定函数管道

同上

> “set fog”或“add light”

把这段翻译了

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
+![Graphics Pipeline](./rendering_pipeline.png)
 
-OpenGL 2.0 introduced the concept of programmable pipeline. In this model, the different steps that compose the graphics pipeline can be controlled or programmed by using a set of specific programs called shaders. The following picture depicts a simplified version of the OpenGL programmable pipeline:
+但在OPENGL 2.0版本,其引入了可编程流水线渲染的概念.在这个模型中,组成图形管道的不同步骤可以通过一组称为“着色器”的特定程序来控制或编程.(可编程自定义渲染过程,而不是只能使用固定设置)下图描述了OpenGL可编程管道的简化版本:
```
> 可编程流水线渲染

可编程渲染管线

> 图形管道

渲染管线

> (可编程自定义渲染过程,而不是只能使用固定设置)

原文有这一段吗？

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
+但在OPENGL 2.0版本,其引入了可编程流水线渲染的概念.在这个模型中,组成图形管道的不同步骤可以通过一组称为“着色器”的特定程序来控制或编程.(可编程自定义渲染过程,而不是只能使用固定设置)下图描述了OpenGL可编程管道的简化版本:
 
-![Programmable pipeline](rendering_pipeline_2.png)
+![Programmable pipeline](./rendering_pipeline_2.png)
```
[]内的是图片描述，需要翻译

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
+顶点着色器会处理这些顶点,其主要目的是计算每个顶点在屏幕空间中的投影位置.该着色器还可以生成与颜色或纹理相关的其他输出,但其主要目标是将顶点投影到屏幕空间,即生成点.
 
-The geometry processing stage connects the vertices that are transformed by the vertex shader to form triangles. It does so by taking into consideration the order in which the vertices were stored and grouping them using different models. Why triangles? A triangle is like the basic work unit for graphic cards. It’s a simple geometric shape that can be combined and transformed to construct complex ...
+几何体的处理阶段:连接顶点着色器变换的顶点以形成三角形.它是通过判断顶点的存储顺序并使用不同的模型对它们进行分组来实现的.那为什么是三角形呢?三角形类似于图形卡的基本工作单元.它是一个简单的几何形状,可以组合和转换来构建复杂的三维场景.此阶段还可以使用特定着色器对顶点进行分组.
```
> 变换

转换

> 形成

生成

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
+片段着色器在片段处理阶段会使用这些片段来生成像素,为它们指定写入帧缓冲区的最终颜色.帧缓冲区是图形管道的最终结果.它保存应绘制到屏幕上的每一个像素的值.
 
-Keep in mind that 3D cards are designed to parallelize all the operations described above. The input data can be processed in parallel in order to generate the final scene.
+请记住,3D加速卡专门设计来用于并行上述所有操作.为了生成最终场景,可以并行处理输入数据.(非常多的线程共同计算,参考英伟达的CUDA芯片,几千个核心共同处理,这是专为渲染图形所设计的硬件方案)
```
括号里的东西全删了

Qubik65536 commented on chapter04/chapter4.md:
````diff
@@ -37,9 +37,9 @@ void main()
 ```
 
-The first line is a directive that states the version of the GLSL language we are using. The following table relates the GLSL version, the OpenGL that matches that version and the directive to use \(Wikipedia: [https://en.wikipedia.org/wiki/OpenGL\_Shading\_Language\#Versions](https://en.wikipedia.org/wiki/OpenGL_Shading_Language#Versions)\).
+第一行是一个指令,说明了我们使用的GLSL语言的版本.下表则说明了GLSL版本、和与该版本匹配的OpenGL以及要使用的指令.\(维基百科: [https://en.wikipedia.org/wiki/OpenGL\_Shading\_Language\#Versions](https://en.wikipedia.org/wiki/OpenGL_Shading_Language#Versions)\).
````
改成中文链接：https://zh.wikipedia.org/wiki/GLSL

WUTONK committed:
zh_cn chapter5

Initial translation into Chinese, if there are punctuation and other issues, please feedback

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -54,13 +54,13 @@ The first line is a directive that states the version of the GLSL language we ar
 | 4.50 | 4.5 | \#version 450 |
 
-The second line specifies the input format for this shader. Data in an OpenGL buffer can be whatever we want, that is, the language does not force you to pass a specific data structure with a predefined semantic. From the point of view of the shader it is expecting to receive a buffer with data. It can be a position, a position with some additional information or whatever else we want. The vertex ...
+第二行用来指定此着色器的输入格式.OpenGL缓冲区中的数据都可以是我们想要的任何数据,也就是说,该语言不会强制你传递具有预定义语义的特定数据结构.从着色器的角度来看,它希望接收一个包含数据的缓冲区.它可以是一个位置,一个有额外信息的位置,或者我们想要的任何位置.当顶点着色器刚刚接收一个浮点数组,填充缓冲区时,定义将由着色器处理的缓冲区块来定义.
```
> 或者我们想要的任何位置

`or whatever else we want`，我想我没有要说的话了

> 刚刚接收

`is just receving`，仅仅接收到

------

最后这一段完全不是你翻译的意思

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -54,13 +54,13 @@ The first line is a directive that states the version of the GLSL language we ar
+所以,首先,我们需要把这些转化成对我们有意义的东西.在本例中,从位置0开始,我们期望接收一个由3个参数组成的向量坐标\(x, y, z\).
 
-The shader has a main block like any other C program which in this case is very simple. It is just returning the received position in the output variable `gl_Position` without applying any transformation. You now may be wondering why the vector of three attributes has been converted into a vector of four attributes \(vec4\). This is because `gl_Position` is expecting the result in vec4 format sin ...
+着色器有一个主块,就像任何其他的C语言程序一样,在本案例中非常简单,它要做的只是返回输出变量`gl_Position`中接收到的位置,而不应用任何转换.现在你可能想知道为什么三个属性的向量会被转换成四个属性的向量,这到底是怎么实现的\(vec4\)？这是因为`gl_Position`使用的是齐次坐标格式,因此预期结果为vec4格式.也就是说,它需要某种形式的四个变量\(x，y，z，w\)s,其中w表示一个额外的维度.那为什么要添加另一个维度呢?在后面的章节中,你将看到我们需要做的大多数操作都是基于向量和矩阵的.如果没有那个额外的维度,这些操作中的一些就不能合并.例如,如果没有这个额外的维度,将不能结合旋转和平移操作.\(如果你想了解更多这个额外的维度，以允许结合仿射和线性变换。你可以通过阅读Fletcher Dunn和Ian Parberry大佬的优秀著作《3D Math Primer  ...
```
> 现在你可能想知道为什么三个属性的向量会被转换成四个属性的向量,这到底是怎么实现的\(vec4\)？

哪里来的“这到底是怎么实现的”？

> 这些操作中的一些就不能合并

不能被同时使用

> 《3D Math Primer for Graphics and Game Development

"》"去哪了？

> 阅读Fletcher Dunn和Ian Parberry大佬的优秀著作

谁给你添油加醋的权利的？

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -54,13 +54,13 @@ The first line is a directive that states the version of the GLSL language we ar
+着色器有一个主块,就像任何其他的C语言程序一样,在本案例中非常简单,它要做的只是返回输出变量`gl_Position`中接收到的位置,而不应用任何转换.现在你可能想知道为什么三个属性的向量会被转换成四个属性的向量,这到底是怎么实现的\(vec4\)？这是因为`gl_Position`使用的是齐次坐标格式,因此预期结果为vec4格式.也就是说,它需要某种形式的四个变量\(x，y，z，w\)s,其中w表示一个额外的维度.那为什么要添加另一个维度呢?在后面的章节中,你将看到我们需要做的大多数操作都是基于向量和矩阵的.如果没有那个额外的维度,这些操作中的一些就不能合并.例如,如果没有这个额外的维度,将不能结合旋转和平移操作.\(如果你想了解更多这个额外的维度，以允许结合仿射和线性变换。你可以通过阅读Fletcher Dunn和Ian Parberry大佬的优秀著作《3D Math Primer  ...
 
-Let us now have a look at our first fragment shader. We will create a file named “`fragment.fs`” \(the extension is for Fragment Shader\) under the resources directory with the following content:
+现在让我们看看我们的第一个片段着色器.我们将在resources目录下创建一个名为“`fragment.fs`”的方法\(扩展名是 Fragment Shader\)位于resources目录下,包含以下内容:
```
> 位于resources目录下,包含以下内容:

句号。在resources目录下,包含以下内容:

Qubik65536 commented on chapter04/chapter4.md:
````diff
@@ -167,21 +167,21 @@ public class ShaderProgram {
 ```
 
-The constructor of the `ShaderProgram` creates a new program in OpenGL and provides methods to add vertex and fragment shaders. Those shaders are compiled and attached to the OpenGL program. When all shaders are attached the link method should be invoked which links all the code and verifies that everything has been done correctly.
+这个叫`ShaderProgram`的构造函数在OpenGL中创建了一个新程序,并提供了添加顶点和片段着色器的方法.这些着色器被编译并附加到OpenGL程序.当附加所有着色器时,应该调用link方法,该方法负责链接所有代码,并验证所有操作是否正确.
````
> 这些着色器被编译并附加到OpenGL程序

这些着色器会被编译并附加到OpenGL程序

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -167,21 +167,21 @@ public class ShaderProgram {
+当链接着色器程序后,可以释放已编译的顶点和片段着色器\(通过调用 `glDetachShader`\).
 
-Regarding verification, this is done through the `glValidateProgram` call. This method is used mainly for debugging purposes, and it should be removed when your game reaches production stage. This method tries to validate if the shader is correct given the **current OpenGL state**. This means, that validation may fail in some cases even if the shader is correct, due to the fact that the current s ...
+此方法主要用于调试目的,当游戏进入生产阶段时应将其删除.此方法尝试验证给定**当前OpenGL状态**的着色器是否正确.这意味着,在某些情况下,即使着色器是正确的,验证也可能失败,因为当前状态不够完整,无法运行着色器\(某些数据可能尚未上载\),因此,我们只需将错误消息打印到标准错误输出就好,而不是认为运行失败.
```
> Regarding verification, this is done through the `glValidateProgram` call.

去哪了？

> 因为当前状态不够完整,无法运行着色器

因为当前状态不够完整运行着色器

Qubik65536 commented on chapter04/chapter4.md:
````diff
@@ -214,9 +214,9 @@ public void init() throws Exception {
 ```
 
-We have created a utility class which by now provides a method to retrieve the contents of a file from the class path. This method is used to retrieve the contents of our shaders.
+我们已经创建了一个实用程序类,现在它提供了一个从类路径检索文件内容的方法,此方法用于检索着色器的内容.
````
> 实用程序类

工具类

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -226,43 +226,43 @@ float[] vertices = new float[]{
+顶点数组对象\(VAOs\)是包含一个或多个VBO(通常称为属性列表)的对象.每个属性列表可以保存一种类型的数据:位置、颜色、纹理等,这样你就可以可以随意在每个插槽中存储任何你想要的数据.
 
-A VAO is like a wrapper that groups a set of definitions for the data that is going to be stored in the graphics card. When we create a VAO we get an identifier. We use that identifier to render it and the elements it contains using the definitions we specified during its creation.
+VAO就像一个包装器,它为将要存储在图形卡中的数据分组为一组一组的定义,当我们创建一个VAO时,我们得到一个标识符.我们使用该标识符来呈现它和它包含的元素,使用我们在创建期间指定的定义.
```
> 我们得到一个标识符

我们将会得到一个标识符

> 我们使用该标识符来呈现它和它包含的元素,使用我们在创建期间指定的定义.

我们使用该标识符和我们在创建期间指定的定义来呈现它和它包含的元素

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -226,43 +226,43 @@ float[] vertices = new float[]{
+VAO就像一个包装器,它为将要存储在图形卡中的数据分组为一组一组的定义,当我们创建一个VAO时,我们得到一个标识符.我们使用该标识符来呈现它和它包含的元素,使用我们在创建期间指定的定义.
 
-So let us continue coding our example. The first thing that we must do is to store our array of floats into a `FloatBuffer`. This is mainly due to the fact that we must interface with the OpenGL library, which is C-based, so we must transform our array of floats into something that can be managed by the library.
+让我们继续编写我们的示例.我们必须做的第一件事是将浮点数组存储到`FloatBuffer`中.这主要是因为我们必须与基于C语言的OpenGL库接口,所以我们必须将我们的float数组转换成库可以管理的东西.
```
> 这主要是因为我们必须与基于C语言的OpenGL库接口

你自己觉得通顺吗

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -226,43 +226,43 @@ float[] vertices = new float[]{
+我们使用`MemoryUtil`类在堆外内存中创建缓冲区,以便OpenGL库可以访问它.在我们存储了数据\(使用put方法\)之后,我们还需要使用flip方法\(也就是说,我们已经完成了对它的写入\)将缓冲区的位置重置为0位置.请记住,Java对象是在称为堆的空间中分配的.堆是JVM进程内存中保留的大量内存空间.存储在堆中的内存不能被本机代码访问\(JNI,允许从Java调用本机代码的机制不允许这样\).在Java和本机代码之间共享内存数据的唯一方法是直接在Java中分配内存.
 
-If you come from previous versions of LWJGL it's important to stress out a few topics. You may have noticed that we do not use the utility class `BufferUtils` to create the buffers. Instead we use the `MemoryUtil` class. This is due to the fact that `BufferUtils` was not very efficient, and has been mantained only for backwards compatibility. Instead, LWJGL 3 proposes two methods for buffer manag ...
+如果你是LWJGL的早期版本使用者,那么强调几个主题是很重要的.你可能已经注意到,我们没有使用实用程序类`BufferUtils`来创建缓冲区.相反,我们使用`MemoryUtil`类.这是由于`BufferUtils`的效率不高,而且只是为向后兼容而设计的.相反,LWJGL 3则提出了两种缓冲区管理方法:
```
> 相反,LWJGL 3则提出了两种缓冲区管理方法:

为此,LWJGL...

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -271,40 +271,40 @@ glBufferData(GL_ARRAY_BUFFER, verticesBuffer, GL_STATIC_DRAW);
-* stride: Specifies the byte offset between consecutive generic vertex attributes. \(We will explain it later\).
-* offset: Specifies an offset to the first component in the buffer.
+* 索引:指定着色器需要的数据的位置.
+* 大小:指定每个顶点的组件数属性\(from 1 to 4\).在这个例子中,我们传递的是三维坐标,所以它应该是3.
```
> \(from 1 to 4\).

翻译了啊

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -271,40 +271,40 @@ glBufferData(GL_ARRAY_BUFFER, verticesBuffer, GL_STATIC_DRAW);
 glBindBuffer(GL_ARRAY_BUFFER, 0);
 
-// Unbind the VAO
+// 关闭VAO
```
> 关闭VAO

解除VAO绑定

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -1,11 +1,11 @@
+# 有关渲染的详细信息
 
-In this chapter we will continue talking about how OpenGL renders things. In order to tidy up our code a little bit let’s create a new class called Mesh which, taking as an input an array of positions, creates the VBO and VAO objects needed to load that model into the graphics card.
+建一个名为`Mesh`的新的类,它将用一个位置数组作为输入,创建该模型加载到显卡中所需的VBO和VAO对象.
```
> In this chapter we will continue talking about how OpenGL renders things. In order to tidy up our code a little bit

都去哪了？

Qubik65536 commented on chapter05/chapter5.md:
````diff
@@ -53,18 +53,18 @@ public class Mesh {
 ```
 
-We will create our `Mesh` instance in our `DummyGame` class, removing the VAO and VBO code from `Renderer` `init` method. Our render method in the `Renderer` class will accept also a Mesh instance to render. The `cleanup` method will also be simplified since the `Mesh` class already provides one for freeing VAO and VBO resources.
+我们将在`DummyGame`类中创建`Mesh`实例,从`Renderer``init`方法中删除VAO和VBO代码.`Renderer`类中的render方法也将接受要渲染的网格实例.`cleanup`方法也将得到简化,因为`Mesh`类已经提供了释放VAO和VBO资源的方法.
````
> `Renderer``init`

原来中间有空格为什么删了？

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -94,17 +94,20 @@ public void cleanup() {
+![Quad coordinates](./quad_coordinates.png)
 
-As you can see each of the two triangles is composed of three vertices. The first one formed by the vertices V1, V2 and V4 \(the orange one\) and the second one formed by the vertices V4, V2 and V3 \(the green one\). Vertices are specified in a counter-clockwise order, so the float array to be passed will be \[V1, V2, V4, V4, V2, V3\]. Thus, the init method in our `DummyGame` class will be:
+
```
107行的东西为什么在108行上？

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -226,43 +226,43 @@ float[] vertices = new float[]{
+下图描绘了我们坐标系中的三角形:
 
-![Triangle](triangle_coordinates.png)
+![Triangle](./triangle_coordinates.png)
```
[]内的是图片描述，需要翻译

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -342,19 +342,18 @@ public void cleanup() {
-![Triangle game](triangle_window.png)
-
-Our first triangle! You may think that this will not make it into the top ten game list, and you will be totally right. You may also think that this has been too much work for drawing a boring triangle. But keep in mind that we are introducing key concepts and preparing the base infrastructure to do more complex things. Please be patient and continue reading.
+![Triangle game](./triangle_window.png)
```
[]内的是图片描述，需要翻译

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -122,34 +125,37 @@ public void init() throws Exception {
-![Quad rendered](quad_rendered.png)
+现在，你应该看到这样渲染的四边形:
+![Quad rendered](./quad_rendered.png)
+
```
[]内的是图片描述，需要翻译

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -122,34 +125,37 @@ public void init() throws Exception {
-Now you should see a quad rendered like this:
 
-![Quad rendered](quad_rendered.png)
+现在，你应该看到这样渲染的四边形:
```
空行去哪了？

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -94,17 +94,20 @@ public void cleanup() {
+我们的`Mesh`通过将位置数组长度除以 3 来计算顶点的数量.(因为我们传递的是X、Y和Z坐标).现在我们可以渲染更复杂的形状,让我们尝试渲染更复杂的形状————渲染一个四边形! 四边形可以通过使用两个三角形来构造,如下图所示:
 
-![Quad coordinates](quad_coordinates.png)
+![Quad coordinates](./quad_coordinates.png)
```
[]内的是图片描述，需要翻译

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -122,34 +125,37 @@ public void init() throws Exception {
+我们的工作结束了吗?不幸的是,并没有.上面的代码仍然着存在一些问题.我们在重复坐标来表示四边形.经过两次V2和V4坐标,有了这个复杂性小的外形,这看起来还没什么,但是想象一下一个更复杂的3D模型.我们会多次重复坐标.请记住,现在我们只是使用三个浮点数来表示顶点的位置.但稍后我们将需要更多的数据来表示纹理等.还要考虑到,在更复杂的形状中,三角形之间共享的顶点数可能会更多,如下图所示(其中一个顶点甚至可以在六个三角形之间共享):
 
-![Dolphin](dolphin.png)
+![Dolphin](./dolphin.png)
```
[]内的是图片描述，需要翻译

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -122,34 +125,37 @@ public void init() throws Exception {
 | 0 | 1 | 2 | 3 |
 
-Then we specify the order in which those vertices should be drawn by referring to their position:
+
```
为什么多了个空行

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -122,34 +125,37 @@ public void init() throws Exception {
 
-So we need to modify our `Mesh` class to accept another parameter, an array of indices, and now the number of vertices to draw will be the length of that indices array.
 
+因此,我们需要修改`Mesh`类,以接受另一个参数 -- 一个索引数组.现在要绘制的顶点数将是该索引数组的数.
```
为什么少了个空行
------
> 将是该索引数组的数

be the length of that indices array，该索引数组的长度

Qubik65536 commented on chapter05/chapter5.md:
````diff
@@ -122,34 +125,37 @@ public void init() throws Exception {
 ```
 
-After we have created our VBO that stores the positions, we need to create another VBO which will hold the indices. So we rename the identifier that holds the identifier for the positions VBO and create a new one for the index VBO \(`idxVboId`\). The process of creating that VBO is similar but the type is now `GL_ELEMENT_ARRAY_BUFFER`.
+创建存储位置的VBO后,需要创建另一个保存索引的VBO.因此,我们需要重命名保存位置,和VBO标识符的标识符,并为索引VBO\(`idxVboId`\)创建一个新的标识符.这和创建VBO的过程类似,不过类型现在是`GL_ELEMENT_ARRAY_BUFFER`.
````
> 因此,我们需要重命名保存位置,和VBO标识符的标识符,并为索引VBO\(`idxVboId`\)创建一个新的标识符.

> So we rename the identifier that holds the identifier for the positions VBO and create a new one for the index VBO \(`idxVboId`\).

因此，我们将重命名保存着存储位置的VBO的标识符的标识符，然后为索引VBO\(`idxVboId`\)创建一个新的标识符

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -160,45 +166,52 @@ glBufferData(GL_ELEMENT_ARRAY_BUFFER, indicesBuffer, GL_STATIC_DRAW);
+因为我们要处理整数,所以需要创建一个`IntBuffer`(整数类型)而不是`FloatBuffer`(浮点类型)
 
-And that’s it. The VAO will contain now two VBOs, one for positions and another one that will hold the indices and that will be used for rendering. Our cleanup method in our `Mesh` class must take into consideration that there is another VBO to free.
+
```
为什么又多了一个空行

Qubik65536 commented on chapter05/chapter5.md:
````diff
@@ -160,45 +166,52 @@ glBufferData(GL_ELEMENT_ARRAY_BUFFER, indicesBuffer, GL_STATIC_DRAW);
 ```
 
-Finally, we need to modify our drawing call that used the `glDrawArrays` method:
+
````
为什么又多了一个空行

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -160,45 +166,52 @@ glBufferData(GL_ELEMENT_ARRAY_BUFFER, indicesBuffer, GL_STATIC_DRAW);
-* type: Specifies the type of value in the indices data. In this case we are using integers.
-* indices: Specifies the offset to apply to the indices data to start rendering.
+该方法的参数为:
+
```
为什么又多了一个空行

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -160,45 +166,52 @@ glBufferData(GL_ELEMENT_ARRAY_BUFFER, indicesBuffer, GL_STATIC_DRAW);
+该方法的参数为:
+
+* 模式:指定用于渲染的基本体,在本示例中为三角形.这里没有变化.
+
```
为什么又多了一个空行

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -160,45 +166,52 @@ glBufferData(GL_ELEMENT_ARRAY_BUFFER, indicesBuffer, GL_STATIC_DRAW);
 
-And now we can use our newer and much more efficient method of drawing complex models by just specifying the indices.
+* 类型:指定索引数据中的值的类型,在这种情况下,我们使用整数.
+
```
为什么又多了一个空行

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -160,45 +166,52 @@ glBufferData(GL_ELEMENT_ARRAY_BUFFER, indicesBuffer, GL_STATIC_DRAW);
+* 计数:指定要呈现的元素的数量.
 
-And now we can use our newer and much more efficient method of drawing complex models by just specifying the indices.
+* 类型:指定索引数据中的值的类型,在这种情况下,我们使用整数.
```
为什么又多了一个空行

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -160,45 +166,52 @@ glBufferData(GL_ELEMENT_ARRAY_BUFFER, indicesBuffer, GL_STATIC_DRAW);
+* 类型:指定索引数据中的值的类型,在这种情况下,我们使用整数.
+
+* 索引:指定一个offset量并应用在索引数据上才能开始转换.
+
```
为什么又多了一个空行

Qubik65536 commented on chapter05/chapter5.md:
````diff
@@ -253,8 +267,7 @@ void main()
 ```
-
-And now our fragment shader receives as an input the colour processed by our vertex shader and uses it to generate the colour.
+现在,我们的片段着色器接收顶点着色器处理的颜色作为输入,并使用它来生成颜色.
````
为什么又少了一个空行

Qubik65536 commented on chapter05/chapter5.md:
````diff
@@ -267,8 +280,7 @@ void main()
 ```
-
- We can now pass an array of colours like this to our `Mesh` class in order to add some colour to our quad.
+我们现在可以将这样的颜色数组传递给`Mesh`类,以便为我们的四边形添加一些颜色.
````
为什么又少了一个空行

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -278,8 +290,6 @@ float[] colours = new float[]{
-And we will get a fancy coloured quad like this.
-
-![Coloured quad](coloured_quad.png)
+然后我们将会得到一个像这样的花式彩色四边形！:
```
为什么又少了一个空行

Qubik65536 commented on chapter05/chapter5.md:
```diff
@@ -278,8 +290,6 @@ float[] colours = new float[]{
-
-![Coloured quad](coloured_quad.png)
+然后我们将会得到一个像这样的花式彩色四边形！:
+![Coloured quad](./coloured_quad.png)
```
为什么又少了一个空行

------

[]内的是图片描述，需要翻译

Qubik65536 reviewed (changes requested):
大量的问题重复出现
- **为什么标题和描述没有包含第五章**
- **为什么又多了一个空行**
- **为什么又少了一个空行**
- 语句不通顺
- 专业术语不正确

WUTONK committed:
zh_cn chapter4

Fixed problems with computer terminology

PencilCore commented:
我求求你了，在使用半角字符时，请在句子和句子间加空格

PencilCore commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
+渲染管线由以下步骤组成：
 
-![Graphics Pipeline](rendering_pipeline.png)
+![图形管线](./rendering_pipeline.png)
```
？渲染管线

Qubik65536 replied:
`渲染管线由以下步骤组成` 刚刚改过

PencilCore commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
+![图形管线](./rendering_pipeline.png)
 
-OpenGL 2.0 introduced the concept of programmable pipeline. In this model, the different steps that compose the graphics pipeline can be controlled or programmed by using a set of specific programs called shaders. The following picture depicts a simplified version of the OpenGL programmable pipeline:
+但在OPENGL 2.0版本,其引入了可编程渲染管线的概念.在这个模型中,组成图形管道的不同步骤可以通过一组称为“着色器”的特定程序来控制或编程.下图描述了OpenGL可编程管道的简化版本:
```
最后一句话，管道—管线

Qubik65536 commented:
> 我求求你了，在使用半角字符时，请在句子和句子间加空格

ICU欢迎您

PencilCore commented on chapter04/chapter4.md:
```diff
@@ -1,30 +1,30 @@
+![可编程渲染管线](./rendering_pipeline_2.png)
 
-The rendering starts taking as its input a list of vertices in the form of Vertex Buffers. But, what is a vertex? A vertex is a data structure that describes a point in 2D or 3D space. And how do you describe a point in a 3D space? By specifying its x, y and z coordinates. And what is a Vertex Buffer? A Vertex Buffer is another data structure that packs all the vertices that need to be rendered,  ...
+渲染开始将顶点缓冲区形式的顶点列表作为其输入.但,什么是顶点?顶点是描述二维或三维空间中的点的数据结构.你该如何描述三维空间中的一个点?————通过指定其x、y和z坐标.那什么是顶点缓冲区?顶点缓冲区是另一种数据结构,它通过使用顶点数组,打包需要渲染的所有顶点,并使这些信息可供图形管道中的着色器使用.
```
破折号（半角）= -

PencilCore replied:
x, y 和 z 坐标？

PencilCore replied:
最后一句话老问题

PencilCore commented on chapter04/chapter4.md:
```diff
@@ -54,13 +54,13 @@ The first line is a directive that states the version of the GLSL language we ar
+所以,首先,我们需要把这些转化成对我们有意义的东西.在本例中,从位置0开始,我们期望接收一个由3个参数组成的向量坐标\(x, y, z\).
 
-The shader has a main block like any other C program which in this case is very simple. It is just returning the received position in the output variable `gl_Position` without applying any transformation. You now may be wondering why the vector of three attributes has been converted into a vector of four attributes \(vec4\). This is because `gl_Position` is expecting the result in vec4 format sin ...
+着色器有一个主块,就像任何其他的C语言程序一样,在本案例中非常简单,它要做的只是返回输出变量`gl_Position`中接收到的位置,而不应用任何转换.现在你可能想知道为什么三个属性的向量会被转换成四个属性的向量\(vec4\)？这是因为`gl_Position`使用的是齐次坐标格式,因此预期结果为vec4格式.也就是说,它需要某种形式的四个变量\(x，y，z，w\)s,其中w表示一个额外的维度.那为什么要添加另一个维度呢?在后面的章节中,你将看到我们需要做的大多数操作都是基于向量和矩阵的.如果没有那个额外的维度,这些操作中的一些就不能合并.例如,如果没有这个额外的维度,将不能结合旋转和平移操作.\(如果你想了解更多这个额外的维度，以允许结合仿射和线性变换。你可以通过阅读Fletcher Dunn和Ian Parberry的优秀著作《3D Math Primer for Graphics ...
```
不要用书名号，用引号

PencilCore replied:
英文书籍没有用书名号的，都用斜体或引号

Qubik65536 replied:
这个可以不用改，中文文章

> 不要用书名号，用引号

PencilCore replied:
> 这个可以不用改，中文文章
> 
> 
> 
> > 不要用书名号，用引号

https://www.zhihu.com/question/33271098，请参考

Qubik65536 replied:
书名号倒是改了啊！

PencilCore reviewed (commented):
差不多了，这些参考下，然后让cubik并掉吧

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -1,28 +1,28 @@
 顶点着色器会处理这些顶点,其主要目的是计算每个顶点在屏幕空间中的投影位置.该着色器还可以生成与颜色或纹理相关的其他输出,但其主要目标是将顶点投影到屏幕空间,即生成点.
 
-几何体的处理阶段:连接顶点着色器变换的顶点以形成三角形.它是通过判断顶点的存储顺序并使用不同的模型对它们进行分组来实现的.那为什么是三角形呢?三角形类似于图形卡的基本工作单元.它是一个简单的几何形状,可以组合和转换来构建复杂的三维场景.此阶段还可以使用特定着色器对顶点进行分组.
+几何体的处理阶段:连接顶点着色器转换的顶点以生成三角形.它是通过判断顶点的存储顺序并使用不同的模型对它们进行分组来实现的.那为什么是三角形呢?三角形类似于图形卡的基本工作单元.它是一个简单的几何形状,可以组合和转换来构建复杂的三维场景.此阶段还可以使用特定着色器对顶点进行分组.
```
连接顶点着色器转换顶点以生成三角形

Qubik65536 commented on chapter04/chapter4.md:
````diff
@@ -37,7 +37,7 @@ void main()
 ```
 
-第一行是一个指令,说明了我们使用的GLSL语言的版本.下表则说明了GLSL版本、和与该版本匹配的OpenGL以及要使用的指令.\(维基百科: [https://en.wikipedia.org/wiki/OpenGL\_Shading\_Language\#Versions](https://en.wikipedia.org/wiki/OpenGL_Shading_Language#Versions)\).
+第一行是一个指令,说明了我们使用的GLSL语言的版本.下表则说明了GLSL版本、和与该版本匹配的OpenGL以及要使用的指令.\(维基百科: [https://en.wikipedia.org/wiki/OpenGL\_Shading\_Language\#Versions](https://zh.wikipedia.org/wiki/GLSL)\).
````
`[https://en.wikipedia.org/wiki/OpenGL\_Shading\_Language\#Versions]` 里的东西为什么给改了？

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -54,13 +54,13 @@ void main()
+着色器有一个主块,就像任何其他的C语言程序一样,在本案例中非常简单,它要做的只是返回输出变量`gl_Position`中接收到的位置,而不应用任何转换.现在你可能想知道为什么三个属性的向量会被转换成四个属性的向量\(vec4\)？这是因为`gl_Position`使用的是齐次坐标格式,因此预期结果为vec4格式.也就是说,它需要某种形式的四个变量\(x，y，z，w\)s,其中w表示一个额外的维度.那为什么要添加另一个维度呢?在后面的章节中,你将看到我们需要做的大多数操作都是基于向量和矩阵的.如果没有那个额外的维度,这些操作中的一些就不能合并.例如,如果没有这个额外的维度,将不能结合旋转和平移操作.\(如果你想了解更多这个额外的维度，以允许结合仿射和线性变换。你可以通过阅读Fletcher Dunn和Ian Parberry的优秀著作《3D Math Primer for Graphics ...
 
-现在让我们看看我们的第一个片段着色器.我们将在resources目录下创建一个名为“`fragment.fs`”的方法\(扩展名是 Fragment Shader\)位于resources目录下,包含以下内容:
+现在让我们看看我们的第一个片段着色器.我们将在resources目录下创建一个名为“`fragment.fs`”的方法\(扩展名是 Fragment Shader\).位于resources目录下,包含以下内容:
```
在resources目录下,包含以下内容:

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -167,11 +167,11 @@ public class ShaderProgram {
 当链接着色器程序后,可以释放已编译的顶点和片段着色器\(通过调用 `glDetachShader`\).
 
-此方法主要用于调试目的,当游戏进入生产阶段时应将其删除.此方法尝试验证给定**当前OpenGL状态**的着色器是否正确.这意味着,在某些情况下,即使着色器是正确的,验证也可能失败,因为当前状态不够完整,无法运行着色器\(某些数据可能尚未上载\),因此,我们只需将错误消息打印到标准错误输出就好,而不是认为运行失败.
+关于验证,这是通过`glValidateProgram`调用完成的.此方法主要用于调试目的,当游戏进入生产阶段时应将其删除.此方法尝试验证给定**当前OpenGL状态**的着色器是否正确.这意味着,在某些情况下,即使着色器是正确的,验证也可能失败,因为当前状态不够完整运行着色器\(某些数据可能尚未上载\),因此,我们只需将错误消息打印到标准错误输出就好,而不是认为运行失败.
```
> 关于验证,这是通过`glValidateProgram`调用完成的

关于验证,这是通过调用`glValidateProgram`完成的

Qubik65536 reviewed (changes requested):
还有该改的东西没改

Qubik65536 commented:
第五章完全没动？

WUTONK committed:
zh_cn chapter4

Fix errors

WUTONK committed:
zh_cn chapter4

fix some error

Qubik65536 commented on chapter04/chapter4.md:
````diff
@@ -37,7 +37,8 @@ void main()
 ```
 
-第一行是一个指令,说明了我们使用的GLSL语言的版本.下表则说明了GLSL版本、和与该版本匹配的OpenGL以及要使用的指令.\(维基百科: [https://en.wikipedia.org/wiki/OpenGL\_Shading\_Language\#Versions](https://zh.wikipedia.org/wiki/GLSL)\).
+第一行是一个指令,说明了我们使用的GLSL语言的版本.下表则说明了GLSL版本、和与该版本匹配的OpenGL以及要使用的指令.\(维基百科: [https://zh.wikipedia.org/wiki/OpenGL\_Shading\_Language\#Versions](https://zh.wikipedia.org/wiki/GLSL)\).
````
你确定中文维基百科链接是这个？你把后面那一段复制粘贴很难？

Qubik65536 commented on chapter04/chapter4.md:
```diff
@@ -1,26 +1,26 @@
 渲染管线由以下步骤组成：
 
-![图形管线](./rendering_pipeline.png)
+![渲染管线由以下步骤组成](./rendering_pipeline.png)
```
`![渲染管线](./rendering_pipeline.png)`

Qubik65536 commented:
项目暂停，待恢复之后重新Open

KevinZonda reviewed (commented):
This work is amazing! @Cubik65536

Qubik65536 commented:
> This work is amazing! @Cubik65536

Do you want to join this ~~垃圾~~ studio?

KevinZonda commented:
> > This work is amazing! @Cubik65536
> 
> Do you want to join this ~垃圾~ studio?

Yeah, pls!
